matrix/basic_matrix.py

```diff
@@ -144,15 +151,51 @@ def diagonal(self):
             result.append(self.ls_entries[i][i])
         return result
 
+    def LU_decomp(self):
+        """
+        Returns: the LU decomposition of the self.ls_entries matrix
+            L, U where [A] = [L][U]
+            L: <Matrix> m x m lower triangular matrix
+            U: <Matrix> upper triangular matrix
+        """
+        L = self.identity(self.len_row).ls_entries  # L is square matrix
+        U = [[0] * self.len_col for x in range(self.len_row)]
+        A = copy.deepcopy(self.ls_entries)
+
+        # Need to pivot first for stability
+        for i in range(self.len_row):
+            max_row_index = np.argmax(abs(np.asarray(A)[i:self.len_row, i])) + i #todo check with Emil if use of np.asarray is allowed
+            A[i], A[max_row_index] = A[max_row_index], A[i]
+
+        for i in range(self.len_row):
+            for j in range(i, self.len_row):  # update matrix after pivoting
+                sum_upper = sum([L[i][k] * U[k][j] for k in range(i)])
+                sum_lower = sum([L[j][k] * U[k][i] for k in range(i)])
+                U[i][j] = A[i][j] - sum_upper
+                L[j][i] = (A[j][i] - sum_lower) / U[i][i] if U[i][i] != 0 else 0
+
+        return Matrix(L), Matrix(U)
```

Comment:
Question on this, I need to remind myself on the math but do the resulting `L`, `U` matricies need to be square matricies as opposed to any shape?

Reply:
For L and U to be  triangular, yes they should be square, which would come from LU decomp on a square matrix.  Technically, you could use the LU method to factor a non-square mxn matrix, where you'd get L as mxm and U as mxn since the method is just gaussian elimination.  

But I wasn't really thinking of doing that, I mostly just wanted to use the LU decomp to get the determinant after.  This code actually only works for square matrices, so thinking I'll add the is_square check at the beginning

Reply:
Hmm in that case lets not even add an LU decomposition method in the Matrix class, or just keep it as a raise not implemented. 
And only add it to the child class of square matrix: https://github.com/KerimovEmil/HigherMathInvestigations/blob/master/matrix/square_matrix.py

With the idea that the more useful matrices like Hermitian matrices will eventually be child classes of these code classes.

Reply:
Makes sense!  Moved into the square matrix class!